buffer.go

```diff
@@ -28,114 +26,71 @@ type Buffer struct {
 	// This field cannot be nil.
 	Serializer Serializer
 
-	buffer unsafe.Pointer
-	join   sync.WaitGroup
+	mutex  sync.Mutex
+	buffer *buffer
 }
 
 // HandleMeasures satisfies the Handler interface.
-func (h *Buffer) HandleMeasures(time time.Time, measures ...Measure) {
+func (b *Buffer) HandleMeasures(time time.Time, measures ...Measure) {
 	if len(measures) == 0 {
 		return
 	}
 
-	size := h.BufferSize
+	size := b.BufferSize
 	if size == 0 {
 		size = 1024
 	}
 
-	c := chunkPool.Get().(*chunk)
-	c.b = h.Serializer.AppendMeasures(c.b[:0], time, measures...)
-	l := int32(len(c.b))
+	chunk := acquireChunk()
+	chunk.reset()
+	chunk.append(b.Serializer, time, measures...)
 
 	// Chunks that are already larger than the maximum buffer size are directly
 	// passed to the serializer, no need to aggregate them in a larger buffer.
-	if int(l) >= size {
-		h.join.Add(1)
-		go h.writeChunk(c)
-		return
-	}
-
-	for {
-		b := h.loadBuffer()
+	if chunk.len() >= size {
+		chunk.writeTo(b.Serializer)
+	} else {
+		for {
+			var flush *buffer
 
-		if b == nil {
-			b = bufferPool.Get().(*buffer)
-			b.init(size)
+			b.mutex.Lock()
 
-			if !h.compareAndSwapBuffer(nil, b) {
-				bufferPool.Put(b)
-				continue
+			if b.buffer == nil {
+				b.buffer = acquireBuffer()
+				b.buffer.reset(size)
 			}
-		}
-
-		b.m.RLock()
-		m := int32(len(b.b))
-		n := atomic.AddInt32(&b.n, l)
 
-		if n >= 0 && n <= m {
-			copy(b.b[n-l:n], c.b)
-			b.m.RUnlock()
-			break
-		}
+			if (b.buffer.len() + chunk.len()) > b.buffer.cap() {
+				b.buffer, flush = nil, b.buffer
+			} else {
+				b.buffer.append(chunk)
+			}
 
-		b.m.RUnlock()
+			b.mutex.Unlock()
 
-		if n < 0 || (n-l) > m {
-			// Only the goroutine that first overflows the buffer size is in
-			// charge of writing the buffer, the others go back and try to
-			// acquire a new one to copy their measures into.
-			continue
-		}
+			if flush == nil {
+				break
+			}
 
-		if h.compareAndSwapBuffer(b, nil) {
-			b.waitReleaseRLock()
-			h.join.Add(1)
-			go h.writeBuffer(b, int(n-l))
+			flush.writeTo(b.Serializer)
+			releaseBuffer(flush)
 		}
 	}
 
-	chunkPool.Put(c)
+	releaseChunk(chunk)
```

Comment:
could we `defer` this after the `acquire`?

Reply:
`defer` is still not the cheapest construct in Go (it often shows up on performance profiles), in a critical code path like this one I'd rather not use it.